src/nnet3/convolution-cudnn.cc

```diff
@@ -0,0 +1,655 @@
+      CuDevice::Instantiate().GetCudnnHandle(),
+      input_desc_,
+      output_desc_,
+      conv_desc_,
+      params_desc_,
+      bwd_filter_algo_,
+      &temp_space_required_backward_filter_));
+}
+#endif
+
+#if HAVE_CUDA == 1
+void ConvolutionComputation::DestroyCudnn() {
+  CUDNN_SAFE_CALL(cudnnDestroyTensorDescriptor(input_desc_));
+  CUDNN_SAFE_CALL(cudnnDestroyTensorDescriptor(output_desc_));
+  CUDNN_SAFE_CALL(cudnnDestroyFilterDescriptor(params_desc_));
+  CUDNN_SAFE_CALL(cudnnDestroyTensorDescriptor(bias_desc_));
+  CUDNN_SAFE_CALL(cudnnDestroyConvolutionDescriptor(conv_desc_));
+  CUDNN_SAFE_CALL(cudnnDestroyActivationDescriptor(activation_desc_));
+}
+#endif
+
+ConvolutionComputation::~ConvolutionComputation() {
+#if HAVE_CUDA == 1
+  if (CuDevice::Instantiate().Enabled() && descriptors_initialized_)
+    DestroyCudnn();
+#endif
+}
+
+
+void ConvolutionComputation::Read(std::istream &is, bool binary) {
+  config_.Read(is, binary);
+#if HAVE_CUDA == 1
+  if (CuDevice::Instantiate().Enabled()) {
+    InitCudnn();
+  }
+#endif
+}
+
+
+void ConvolutionComputation::
+ConvolveForward(const CuMatrixBase<BaseFloat> &input,
+                const CuMatrixBase<BaseFloat> &params,
+                const CuVectorBase<BaseFloat> &bias,
+                CuMatrixBase<BaseFloat> *output) const {
+  const ConvolutionComputationConfig &c = config_;
+  // Check some dimensions.
+  KALDI_ASSERT(
+      input.NumRows() == c.num_images * c.input_image_width &&
+      input.NumCols() == c.input_image_height * c.num_channels_in &&
+      input.Stride() == input.NumCols() &&
+      params.NumRows() == c.num_channels_out &&
+      params.NumCols() == c.num_channels_in * c.filter_height * c.filter_width &&
+      params.Stride() == params.NumCols());
+  KALDI_ASSERT(
+      bias.Dim() == c.num_channels_out &&
+      output->NumRows() == c.num_images * c.output_image_width &&
+      output->NumCols() == c.output_image_height * c.num_channels_out &&
+      output->Stride() == output->NumCols());
+
+#if HAVE_CUDA == 1
+  if (CuDevice::Instantiate().Enabled()) {
+    CuVector<BaseFloat> temp_space(temp_space_required_forward_ /
+                                   sizeof(BaseFloat), kUndefined);
+    CUDNN_SAFE_CALL(cudnnConvolutionBiasActivationForward(
+        CuDevice::Instantiate().GetCudnnHandle(),
+        &ONE,
+        input_desc_,
+        input.Data(),
+        params_desc_,
+        params.Data(),
+        conv_desc_,
+        fwd_algo_,
+        temp_space.Data(),
+        temp_space.Dim() * sizeof(BaseFloat),
+        &ZERO,
+        output_desc_,
+        output->Data(),
+        bias_desc_,
+        bias.Data(),
+        activation_desc_,
+        output_desc_,
+        output->Data()));
+  } else
+#endif
+  {
+    ConvolveForward(input.Mat(), params.Mat(), bias.Vec(),
+                    &(output->Mat()));
+  }
+}
+
+
+void ConvolutionComputation::
+ConvolveForward(const MatrixBase<BaseFloat> &input,
```

Comment:
Is there some reason why we don't make this class own a "ConvolutionComputation" object, and have it defer to the functions like ConvolveForward defined in convolution.h?